archive/100-theorems-list/88_derangements_formula.lean

```diff
@@ -0,0 +1,468 @@
+/-
+Copyright (c) 2021 Henry Swanson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Henry Swanson
+-/
+import data.fintype.basic
+import data.fintype.card
+import data.equiv.basic
+import group_theory.group_action
+
+import tactic.field_simp
+import tactic.ring
+
+/-!
+# Freek № 88 : Derangements Formula
+This file proves Theorem 88 from the [100 Theorems List](https://www.cs.ru.nl/~freek/100/),
+which gives an explicit formula for the number of derangements (permutations without fixed
+points) on `n` elements.
+
+The core of the proof is an equivalence between "derangements on `n+2` elements", and
+an expression involving derangements on `n` and `n+1` elements. This gives a recurrence
+relation `d(n+2) = (n+1) * (d(n) + d(n+1))`, from which a sum-based formula can be derived.
+-/
+
+open equiv fintype
+open_locale big_operators
+
+section definitions
+
+/-- A permutation is a derangement if it has no fixed points. -/
+def is_derangement {α : Type*} (f : perm α) : Prop := (∀ x : α, f x ≠ x)
+
+/-- The set of derangements on `α`. -/
+def derangements (α : Type*) : set (perm α) := {f : perm α | is_derangement f}
+
+end definitions
+
+
+section derangement_lemmas
+
+/-- If `α` is equivalent to `β`, then `derangements α` is equivalent to `derangements β`. -/
+def derangement_congr {α β : Type*} (e : α ≃ β) : (derangements α ≃ derangements β) :=
+begin
+  refine subtype_equiv (perm_congr e) _,
+  intro f,
+  unfold derangements is_derangement,
+  simp,
+  rw ← not_iff_not,
+  push_neg,
+  split,
+  { rintro ⟨x, hx_fixed⟩,
+    use e x,
+    simp [hx_fixed] },
+  { rintro ⟨y, hy_fixed⟩,
+    use e.symm y,
+    apply e.injective,
+    simp [hy_fixed] }
+end
+
+end derangement_lemmas
+
+
+section ignore_swap
+
+variables {α : Type*} [decidable_eq α]
+
+/-- A predicate describing if permutation `f` swaps `a` and `b`.
+    Note that `a` and `b` need not be distinct. -/
+def swaps {α : Type*} (f : perm α) (a b : α) : Prop := f a = b ∧ f b = a
+
+/-- The set of permutations that swap `a` and `b` is equivalent to permutations on all other
+    elements. -/
+def perm_ignore_swap (a b : α) : { f : perm α | swaps f a b } ≃ perm ({a, b}ᶜ : set α) :=
+begin
+  let a_and_b : set α := {a, b},
+
+  -- TODO why can't it infer this is decidable?
+  haveI : decidable_pred a_and_b := by
+  { dsimp [decidable_pred, a_and_b],
+    intro x,
+    apply set.decidable_mem },
+
+  let swap : perm α := equiv.swap a b,
+  let swap' : perm a_and_b := perm.subtype_perm swap (by {
+    intro x,
+
+    simp,
```

Comment:
No non-terminal `simp`: use `simp?` or `squeeze_simp` and replace this with a `simp only [...]`.

Reply:
oh interesting! didn't know that was a general guideline. is that for performance reasons?

Reply:
It improves performance a bit, but it's mostly for maintenance reasons. If someone adds a new simp-lemma to mathlib, suddenly simp might do more (or different) to what it used to do, and the rest of the proof will break. So then someone (who typically doesn't really know this proof) has to come and fix it. And so then you have to go back to an old copy of mathlib, to find what the goal after the `simp` is supposed to look like. You can see why that's annoying.
With `simp only [lem1, lem2, ...]` it's basically guaranteed to always find the same result. (Unless someone changes the statement of `lem1`, but then that person probably knows quite easily how to fix this proof, because it is exactly clear which ingredients `simp` is using (and not using).

Reply:
oh interesting, that makes sense! so i should convert all my non-terminal `simp`s to `simp only`s?

Reply:
yep! also, you can go for `suffices [TACTIC STATE AFTER SIMP], by simpa`, which I quite like